Arm dac+arm pwm (#6165)
qmk/qmk_firmware: docs/feature_audio.md

### PWM (software)
if the DAC pins are unavailable (or the MCU has no usable DAC at all, like STM32F1xx); PWM can be an alternative.
Note that there is currently only one speaker/pin supported.

set in `rules.mk`:
`AUDIO_DRIVER = pwm_software` and in `config.h`
`#define AUDIO_PIN C13` (can be any pin)
to have the selected pin output a pwm signal, generated from a timer callback which toggles the pin in software.

#### Wiring
the usual piezo wiring: red goes to the selected AUDIO_PIN, black goes to ground.

OR if you can chose to drive one piezo with two pins, for example `#define AUDIO_PIN B1`, `#define AUDIO_PIN_ALT B2` in `config.h`, with `#define AUDIO_PIN_ALT_AS_NEGATIVE` - then the red lead could go to B1, the black to B2.

### PWM (hardware)
STM32F1xx have to fall back to using PWM, but can do so in hardware; but again on currently only one speaker/pin.

`AUDIO_DRIVER = pwm_hardware` in `rules.mk`, and in `config.h`:
`#define AUDIO_PIN A8`

> `#define AUDIO_PWM_TIMER 1`
> `#define AUDIO_PWM_TIMERCHANNEL 1`

Author: ```suggestion
`#define AUDIO_PWM_DRIVER PWMD1`
`#define AUDIO_PWM_CHANNEL 1`
```

> (as well as `#define AUDIO_PWM_PINALTERNATE_FUNCTION 42` if you are on STM32F2 or larger)

Author: ```suggestion
(as well as `#define AUDIO_PWM_PAL_MODE 42` if you are on STM32F2 or larger)
```

# Audio

Your keyboard can make sounds! If you've got a Planck, Preonic, or basically any keyboard that allows access to certain PWM-capable pins, you can hook up a simple speaker and make it beep. You can use those beeps to indicate layer transitions, modifiers, special keys, or just to play some funky 8bit tunes.

To activate this feature, add `AUDIO_ENABLE = yes` to your `rules.mk`.

## AVR based boards
Up to two simultaneous audio voices are supported, one driven by timer 1 and another driven by timer 3.  The following pins can be defined as audio outputs in config.h:

the primary voice, with Timer 3 on one of these pins:
`#define C4_AUDIO`
`#define C5_AUDIO`
`#define C6_AUDIO`

and *optionally*, a secondary voice, using Timer 1, on one of these pins:
`#define B5_AUDIO`
`#define B6_AUDIO`
`#define B7_AUDIO`

## ARM based boards
STM32F2 upwards can use the builtin DAC unit, to drive Pins A4 or A5:

> add `AUDIO_DRIVER = dac` to `rules.mk` and set in `config.h` either:

Reviewer: ```suggestion
Most STM32 MCUs have DAC peripherals, with a notable exception of the STM32F1xx series. Generally, the DAC peripheral drives pins A4 or A5. To enable DAC-based audio output on STM32 devices, add `AUDIO_DRIVER = dac` to `rules.mk` and set in `config.h` either:
```

`#define A4_AUDIO`
OR
`#define A5_AUDIO`

STM32F1xx have to fall back to using PWM (on the up side: with any pin you choose),

either set:
`AUDIO_SYSTEM = pwm-pin-alternate` in `rules.mk` and
`#define AUDIO_PIN C13` in `config.h`
to have the selected pin output a pwm signal, generated from a timer callback (e.g. toggled in software)
 OR
`AUDIO_DRIVER PWM_PIN_ALTERNATE` in `rules.mk`
which will use Timer 1 to directly drive pin PA8 through the PWM hardware (TIM1_CH1 = PA8).

> Should you want to use the pwm-hardware on another pin and timer - be ready to dig into the STM32 datasheet and do the neccessary changes to quantum/audio/audio_arm_pwm.c.

Reviewer: If there are changes required to `quantum/audio/audio_arm_pwm.c` in order to change the pin, please parameterise the inputs via `#define` in order to allow people to perform those modifications by editing their own `config.h`, rather than the qmk sources.

## Songs

> There's a couple different sounds that will automatically be enabled without any other configuration:

Reviewer: ```suggestion
There's a couple of different sounds that will automatically be enabled without any other configuration:
```

# Audio

Your keyboard can make sounds! If you've got a spare pin you can hook up a simple speaker and make it beep. You can use those beeps to indicate layer transitions, modifiers, special keys, or just to play some funky 8bit tunes.

To activate this feature, add `AUDIO_ENABLE = yes` to your `rules.mk`.

## AVR based boards
Up to two simultaneous tones can be rendered. with one speaker driven by timer 1 and another driven by timer 3.  The following pins can be configured as audio outputs in `config.h`:

> for the primary speaker, with Timer 3, pick ONE these pins:

Reviewer 2: Still not a fan of the config option here, if its changing it should be to something more consistent.

For example:

```c
#define AUDIO_PIN      C4
#define AUDIO_PIN_ALT  B5
```

Author: that sadly won't work... since it gets expanded to whatever platform-define C4/B5 will be, ifdef/define switches break... of which there are quite a few to disable unused code-paths - the only solution i currently see is `#define AUDIO_PIN_C4` and the like, which can be used in `#if defined(AUDIO_PIN_C4)`

the nice/consistent solution would have been in the makefile (before the gcc-preprocessor), but that was discouraged :'-(

Reviewer 2: Except it does work, and is already used elsewhere in the code base <https://github.com/qmk/qmk_firmware/blob/master/quantum/backlight/backlight_avr.c#L23>.

Author: the chibios cases that give me trouble with these kind of defines would/could be:
- one speaker on A4+Gnd -> need to enable only DACD1
- two separate (and/or different) speakers on A4/A5+Gnd, assigning one as primary, the other as secondary

Author: how about this compromise: audio_chibios_dac_[basic|additive] keep the `define AUDIO_PIN_A4` type defines, while audio_[avr|chibios]_pwm_* use `define AUDIO_PIN X0` (where the direct preprocessor comparison works, because B7&co expand to simple hex/int values)

i'd gladly use one consistent type of defines, but currently don't see a good/better solution for the chibios_dac side :-S

Author: ha! - found a good(?) solution for the dac side: do more of the AUDIO_PIN checks in c-code instead of the preprocessor - see 677ed4a4a90fac2e74045851616a4ca7413df029

with that the configuration is consitent again over all drivers: AUDIO_PIN A4 - no underscore 

thoughts? does this resolve this thread?

Author: marking this sub-thread as resolved - changes are pushed and need another review...

`#define AUDIO_PIN_C4`
`#define AUDIO_PIN_C5`
`#define AUDIO_PIN_C6`

> and *optionally*, a secondary speaker, using Timer 1, on ONE of these pins:

Reviewer 3: I'm not a huge fan of the wording here.   You can use Timer 1 (B pins) as the primary pin.  This is common for Elite C controllers, as it exposes a number of B pins that aren't used. 

Additionally, some controllers don't have the timer 3, IIRC.

Author: since this is in the AVR specific part, how about i mention Atmega32U4 as the MCU?
this might make things clearer/specific

Author: 3223faa reworks the define names a bit to also include/handle AUDIO_PIN_Bx as only/primary output - plus updated documentation.
better this way?

Reviewer 3: It's better, but I'm still not exactly happy with it.  

I think that is because, at least for the atmega32u4, you can use either timer 1 or timer 3, and that neither should be considered primary/secondory. Unless there is a good reason (aside from having both connected). 

As far as I'm aware, both sets are completely valid and there is no issues with using either.

Author: since zvecr also is in favor of `AUDIO_PIN Bx` for AVR, i reverted only those and updated the documentation accordingly - how is this:  3566defcfb3a852b77fb5b11106ecae0d09f61fc ?

Author: marking this sub-thread as resolved - changes are pushed and need another review...
